### PR TITLE
Use Allocator class in buffer manager and add a test for a custom allocator usage

### PR DESCRIPTION
**Pull request by Mytherin**
This PR enlarges the scope of the allocator so it is also used by the buffer manager. In this manner large memory allocations should pass through the allocator class. Small intermediate allocations (such as vectors) do not use this infrastructure yet, however.

This PR also adds a test with example usage of the allocator class, inlined below:

```cpp
struct MyAllocateData : public PrivateAllocatorData {
	MyAllocateData(atomic<idx_t> *memory_counter_p) : memory_counter(memory_counter_p) {
	}

	atomic<idx_t> *memory_counter;
};

data_ptr_t my_allocate_function(PrivateAllocatorData *private_data, idx_t size) {
	auto my_allocate_data = (MyAllocateData *)private_data;
	*my_allocate_data->memory_counter += size;
	return (data_ptr_t)malloc(size);
}

void my_free_function(PrivateAllocatorData *private_data, data_ptr_t pointer, idx_t size) {
	auto my_allocate_data = (MyAllocateData *)private_data;
	*my_allocate_data->memory_counter -= size;
	free(pointer);
}

TEST_CASE("Test using a custom allocator", "[api]") {
	atomic<idx_t> memory_counter;
	memory_counter = 0;

	Allocator my_allocator(my_allocate_function, my_free_function, make_unique<MyAllocateData>(&memory_counter));

	DBConfig config;
	config.allocator = move(my_allocator);
	DuckDB db(nullptr, &config);
	Connection con(db);
	REQUIRE_NO_FAIL(con.Query("CREATE TABLE tbl AS SELECT * FROM range(1000000)"));

	// printf("\nMemory usage: %lld\n", memory_counter.load());

	REQUIRE_NO_FAIL(con.Query("DROP TABLE tbl"));

	// printf("\nMemory usage: %lld\n", memory_counter.load());
}
```

### CODE REVIEW

**Comment by Mytherin** [comment]
CC @mbasmanova

File: src/storage/buffer_manager.cpp [hunk 1]
```diff
@@ -48,7 +49,7 @@ unique_ptr<BufferHandle> BlockHandle::Load(shared_ptr<BlockHandle> &handle) {
 		// this is mainly down to the block manager only having a single pointer into the file
 		// this is relatively easy to fix later on
 		lock_guard<mutex> buffer_lock(buffer_manager.manager_lock);
```
Comment (hawkfish) [inline src/storage/buffer_manager.cpp:51]: Silly question: Does the lock need to cover the `Read`?

**Review by hawkfish — APPROVED** [review]

File: src/storage/buffer_manager.cpp (hunk 1 above)
Comment (Mytherin) [reply src/storage/buffer_manager.cpp:51]: I believe the comment right above that should answer this :)

```
// FIXME: currently we still require a lock for reading blocks from disk
// this is mainly down to the block manager only having a single pointer into the file
// this is relatively easy to fix later on
```

This should indeed be fixed at some point. I am also not sure about file handles there, perhaps on Windows we will need to have multiple file handles for multi-threaded reads as well.

Comment (hawkfish) [reply src/storage/buffer_manager.cpp:51]: Sorry, I didn't see the comment in the diff.

Comment (Mytherin) [reply src/storage/buffer_manager.cpp:51]: No worries, just thought it was amusing :)

File: test/api/test_custom_allocator.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	Allocator my_allocator(my_allocate_function, my_free_function, make_unique<MyAllocateData>(&memory_counter));
+
+	DBConfig config;
+	config.allocator = move(my_allocator);
+	DuckDB db(nullptr, &config);
+	Connection con(db);
+	REQUIRE_NO_FAIL(con.Query("CREATE TABLE tbl AS SELECT * FROM range(1000000)"));
+
+	// printf("\nMemory usage: %lld\n", memory_counter.load());
```
Comment (hannesmuehleisen) [inline test/api/test_custom_allocator.cpp:40]: maybe instead of commented code there could actually be a check on memory_counter here?

Comment (Mytherin) [reply test/api/test_custom_allocator.cpp]: Yes, good point.

**Pull request merged**